Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +254,447 @@ theorem Language.IsContextFree.reverse (L : Language T) :
   fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
 
 end closure_reversal
+
+section lift_sink
+
+/- This section contains only auxiliary constructions that will shorten upcoming proofs of
+closure properties. When combining several grammars together, we usually want to take a sum type of
+their nonterminal types and lift respective nonterminals to this sum type. We subsequently show
+that the resulting grammar preserves derivations of those strings that may contain any terminals but
+only the proper nonterminals. The lifting operation must be injective. The sinking operation must be
+injective on those symbols where it is defined. -/
+
+/-- Lifting `Symbol` to a larger nonterminal type. -/
+def Symbol.map {N₀ N : Type*} (liftN : N₀ → N) : Symbol T N₀ → Symbol T N
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal n => Symbol.nonterminal (liftN n)
+
+/-- Sinking `Symbol` from a larger nonterminal type; may return `none`. -/
+def Symbol.filterMap {N₀ N : Type*} (sinkN : N → Option N₀) : Symbol T N → Option (Symbol T N₀)
+  | Symbol.terminal t => some (Symbol.terminal t)
+  | Symbol.nonterminal n => Option.map Symbol.nonterminal (sinkN n)
+
+/-- Lifting `ContextFreeRule` to a larger nonterminal type. -/
+def ContextFreeRule.lift {N₀ N : Type*} (r : ContextFreeRule T N₀) (liftN : N₀ → N) :
+    ContextFreeRule T N :=
+  ⟨liftN r.input, r.output.map (Symbol.map liftN)⟩
+
+/-- Lifting `ContextFreeGrammar` to a larger nonterminal type. -/
+structure LiftedContextFreeGrammar (T : Type uT) where
```

Comment:
Your docstring sounds like `LiftedContextFreeGrammar` is an operation, or it's really not. Mind clarifying the docstring?

Reply:
Yeah. I should definitely rephrase it as a structure that represents a pair of grammars over the same terminal alphabet such that (...).